src/appleseed/renderer/modeling/bsdf/microfacetnormalmappinghelper.h

```diff
@@ -0,0 +1,550 @@
+        const int                   modes) const;           	// enabled scattering modes
+
+  private:
+	float heaviside(const float a) const
+	{
+		if (a < 0.0f)
+			return 0.0f;
+		else
+			return 1.0f;
+	}
+
+	float adot(const foundation::Vector3f& a, const foundation::Vector3f& b) const
+	{
+		return std::min(std::abs(foundation::dot(a, b)), 1.0f);
+	}
+
+	float cdot(const foundation::Vector3f& a, const foundation::Vector3f& b) const
+	{
+		return std::max(foundation::dot(a, b), 0.0f);
+	}
+
+	foundation::Vector3f build_tangent(const foundation::Vector3f& original_shading_normal, const foundation::Vector3f& perturbed_shading_normal) const
+	{
+		foundation::Basis3f original_shading_basis(original_shading_normal);
+		foundation::Vector3f local_perturbed_shading_normal = original_shading_basis.transform_to_local(perturbed_shading_normal);
+		foundation::Vector3f local_tangent = foundation::normalize(foundation::Vector3f(-local_perturbed_shading_normal.x, 0.0f, -local_perturbed_shading_normal.z)); // appleseed has y up in local space
+		return original_shading_basis.transform_to_parent(local_tangent);
+	}
+
+	float lambda_p(const foundation::Vector3f& wp, const foundation::Vector3f& wi, const foundation::Vector3f& wg) const
+	{
+		foundation::Vector3f wt = build_tangent(wg, wp);
+
+		float wi_dot_wp = cdot(wi, wp);
+		float wi_dot_wt = cdot(wi, wt);
+		float wp_dot_wg = cdot(wp, wg);
+
+		float ap_wi = wi_dot_wp / wp_dot_wg;
+		float at_wi = (wi_dot_wt * std::sqrt(1.0f - std::pow(wp_dot_wg, 2.0f))) / wp_dot_wg;
+
+		return ap_wi / (ap_wi + at_wi);
+	}
+
+	float lambda_t(const foundation::Vector3f& wp, const foundation::Vector3f& wi, const foundation::Vector3f& wg) const
+	{
+		foundation::Vector3f wt = build_tangent(wg, wp);
+
+		float wi_dot_wp = cdot(wi, wp);
+		float wi_dot_wt = cdot(wi, wt);
+		float wp_dot_wg = cdot(wp, wg);
+
+		float ap_wi = wi_dot_wp / wp_dot_wg;
+		float at_wi = (wi_dot_wt * std::sqrt(1.0f - std::pow(wp_dot_wg, 2.0f))) / wp_dot_wg;
+
+		return at_wi / (ap_wi + at_wi);
+	}
+
+	float G1(const foundation::Vector3f& wp, const foundation::Vector3f& w, const foundation::Vector3f& wg, const bool w_wp) const
+	{
+		if (foundation::dot(w, wg) <= 0.0f)
+			return 0.0f;
+
+		float H = 1.0f;
+		foundation::Vector3f wt = build_tangent(wg, wp);
+
+		if (w_wp)
+			H = heaviside(foundation::dot(w, wp));
+		else
+			H = heaviside(foundation::dot(w, wt));
+
+		float w_dot_wg = cdot(w, wg);
+		float wp_dot_wg = cdot(wp, wg);
+		float w_dot_wp = cdot(w, wp);
+		float w_dot_wt = cdot(w, wt);
+
+		float ap_w = w_dot_wp / wp_dot_wg;
+		float at_w = (w_dot_wt * std::sqrt(1.0f - std::pow(wp_dot_wg, 2.0f))) / wp_dot_wg;
+
+		return H * std::min(1.0f, w_dot_wg / (ap_w + at_w));
+	}
+};
+
+
+//
+// MicrofacetNormalMappingHelper class implementation.
+//
+
+template <typename BSDFImpl>
+MicrofacetNormalMappingHelper<BSDFImpl>::MicrofacetNormalMappingHelper(
```

Comment:
I have been investigating this but I am not sure how to do it yet.
Just to explain the design of this class a bit so others can follow. I have designed this class similar to the `bsdfwrapper.h`. If I am not mistaken, this is [policy-based design](https://stackoverflow.com/questions/23868740/c-inherit-class-from-template-parameter). This class is inheriting template parameter which is in this case arbitrary BRDF. All BRDFs in Appleseed are not "static classes" (e.g. [metalBRDF](https://github.com/appleseedhq/appleseed/blob/master/src/appleseed/renderer/modeling/bsdf/metalbrdf.cpp)) therefore in order to use `BRDF::sample()`, `BRDF::evaluate()` and `BRDF::evaluate_pdf()`, base class has to have constructor in which specific BRDF is constructed. Because of this base class can not be "static class". 
So, BSDFWrapper is not "static class" because it has to have constructor for calling the particular template class (BRDF) constructor. This is required because BRDFs are not static classes (e.g. metalBRDF). Similarly is for this class also.
As I have said, I am not sure how to create a "static class" which has to use arbitrary "non-static" class. I would appreciate any hint about this! @dictoon any ideas?